### PR TITLE
bpo-40280: Skip subprocess-based tests on wasm32-emscripten (GH-30615)

### PR DESCRIPTION
**Pull request by tiran**
https://bugs.python.org/issue40280

### CODE REVIEW

File: Misc/NEWS.d/next/Tests/2022-01-14-23-22-41.bpo-40280.nHLWoD.rst [hunk 1]
```diff
@@ -0,0 +1,5 @@
+Add :func:`test.support.requires_subprocess` and
```
Comment (kumaraditya303) [inline Misc/NEWS.d/next/Tests/2022-01-14-23-22-41.bpo-40280.nHLWoD.rst]:
```suggestion
Add :func:`test.support.requires_subprocess` and
```

Comment (kumaraditya303) [reply Misc/NEWS.d/next/Tests/2022-01-14-23-22-41.bpo-40280.nHLWoD.rst]: This should fix doc CI

File: Lib/test/support/__init__.py [hunk 2]
```diff
@@ -43,6 +43,7 @@
     "anticipate_failure", "load_package_tests", "detect_api_mismatch",
     "check__all__", "skip_if_buggy_ucrt_strfptime",
     "check_disallow_instantiation",
+    "requires_subprocess", "requires_fork",
```
Comment (emmatyping) [inline Lib/test/support/__init__.py]: Your test changes for fork are already merged, right?

Comment (tiran) [reply Lib/test/support/__init__.py]: Yes, I split the PR into two PRs and forgot to clean up this line.

File: Lib/lib2to3/tests/test_parser.py [hunk 3]
```diff
@@ -61,6 +61,9 @@ def test_load_grammar_from_pickle(self):
             shutil.rmtree(tmpdir)
 
     @unittest.skipIf(sys.executable is None, 'sys.executable required')
+    @unittest.skipIf(
+        sys.platform == 'emscripten', 'requires working subprocess'
```
Comment (gpshead) [inline Lib/lib2to3/tests/test_parser.py:65]: why not @requires_subprocess ?

File: Lib/test/support/__init__.py [hunk 4]
```diff
@@ -467,15 +468,23 @@ def requires_debug_ranges(reason='requires co_positions / debug_ranges'):
...
 
-# wasm32-emscripten is POSIX-like but does not provide a
-# working fork() or subprocess API.
+# wasm32-emscripten and -wasi are POSIX-like but do not
+# have subprocess or fork support.
 is_emscripten = sys.platform == "emscripten"
+is_wasi = sys.platform == "wasi"
 
-has_fork_support = hasattr(os, "fork") and not is_emscripten
+has_fork_support = hasattr(os, "fork") and not is_emscripten and not is_wasi
```
Comment (gpshead) [inline Lib/test/support/__init__.py:476]: [meta comment] weird, so those platforms have os.fork but it cannot be used? unusual...

File: Lib/lib2to3/tests/test_parser.py (hunk 3 above)
Comment (tiran) [reply Lib/lib2to3/tests/test_parser.py:65]: Because I made a mistake and forgot to change the code :)

File: Lib/test/support/__init__.py (hunk 4 above)
Comment (tiran) [reply Lib/test/support/__init__.py:476]: Emscripten and WASI are weird platforms. They do their best to mimic a Linux-like, mostly POSIX compatible platform with MUSL libc. The libc has a ``fork`` function, but the "operating system" (JavaScript glue code or WASI runtime) don't support the fork call. The syscall always fails with ENOTSUP.

File: Lib/lib2to3/tests/test_parser.py (hunk 3 above)
Comment (tiran) [reply Lib/lib2to3/tests/test_parser.py:65]: Ah no! Now I remember. The lib2to3 package has its own test suite that does not use anything from ``test.support``.

File: Lib/test/support/__init__.py (hunk 4 above)
Comment (gpshead) [reply Lib/test/support/__init__.py:476]: we could ponder disabling it in that situation, but the code could run on a better runtime that supports it (like running on a more recent kernel) so i guess this is just life for now.  thanks!

File: Lib/lib2to3/tests/test_parser.py (hunk 3 above)
Comment (gpshead) [reply Lib/lib2to3/tests/test_parser.py:65]: yuck, but it's a doomed library so no reason to change that now.

**Review by gpshead — APPROVED** [review]

**Review by emmatyping — APPROVED** [review]
This is great!

**Pull request merged**